crates/nu-std/std/testing.nu

```diff
@@ -0,0 +1,250 @@
+            log info $"Running before-all for module ($module.name)"
+            run-test {
+                file: $module.file,
+                before-each: 'let context = {}',
+                after-each: '',
+                test: 'before-all'
+            }
+            | if $in.exit_code == 0 {
+                $in.stdout
+            } else {
+                throw-error {
+                    msg: "Before-all failed"
+                    label: "Failure in test setup"
+                    span: (metadata $in | get span)
+                }
+            }
+        } else {
+            {}
+    }
+
+    let tests = (
+        $module
+        | flatten
+        | rename -c [tests test]
+        | update before-each {|x|
+            if $module.before-each {
+                $"let context = \(($global_context)|merge \(before-each\)\)"
+            } else {
+                $"let context = ($global_context)"
+            }
+        }
+        | update after-each {|x|
+            if $module.after-each {
+                '$context | after-each'
+            } else {
+                ''
+            }
+        }
+        | each {|test|
+            log info $"Running ($test.test) in module ($module.name) with context ($global_context)"
+            $test|insert result {|x|
+                run-test $test
+                | match $in.exit_code {
+                    0 => "pass",
+                    2 => "skip",
+                    _ => "fail",
+                }
+            }
+        }
+    )
+
+    if $module.after-all {
+        log info $"Running after-all for module ($module.name)"
+
+        run-test {
+                file: $module.file,
+                before-each: $"let context = ($global_context)",
+                after-each: '',
+                test: 'after-all'
+        }
+    }
+    return $tests
+}
+
+export def run-tests [
+    --path: path, # Path to look for tests. Default: current directory.
+    --module: string, # Test module to run. Default: all test modules found.
+    --test: string, # Individual test to run. Default: all test command found in the files.
+    --list, # list the selected tests without running them.
+] {
+    let module_search_pattern = ('**' | path join ({
+        stem: ($module | default "*")
+        extension: nu
+    } | path join))
+
+    let path = ($path | default $env.PWD)
+
+    if not ($path | path exists) {
+        throw-error {
+            msg: "directory_not_found"
+            label: "no such directory"
+            span: (metadata $path | get span)
+        }
+    }
+
+    if not ($module | is-empty) {
+        try { ls ($path | path join $module_search_pattern) | null } catch {
+            throw-error {
+                msg: "module_not_found"
+                label: $"no such module in ($path)"
+                span: (metadata $module | get span)
+            }
+        }
+    }
+
+    let modules = (
+        ls ($path | path join $module_search_pattern)
```

Comment:
You can't just iterate through `ls` output here. You have to check the whole directory tree. In the current state, logger tests are not executed, as they are in a separate subdirectory

Reply:
I didn't write this, this is part of the original design

Reply:
Sure, works just fine right now, might be closed